keep-alive/src/main/java/KeepAliveGatt.kt

```diff
@@ -142,8 +159,17 @@ class KeepAliveGatt internal constructor(
         isRunning.compareAndSet(false, true) || return false
 
         scope.launch(CoroutineName("KeepAliveGatt@$bluetoothDevice")) {
+            var connectionAttempts = 1
             while (isActive) {
-                spawnConnection()
+                val successfullyConnected = spawnConnection()
+                onEvent?.invoke(
+                    Event.Disconnected(
+                        DisconnectInfo(
+                            successfullyConnected,
+                            connectionAttempts++
```

Comment:
Just trying to make sure I understand this... `spawnConnection` returns a boolean indicating whether or not the connection attempt was successful. In the case of a non-throwing, unsuccessful attempt, we'll invoke `onEvent` with a `Disconnected` event. Is that right? Is this the same branch of code that runs for _every_ connection attempt (first attempt vs. a reconnection attempt)? If so, I think it's pretty weird for a consumer to be notified with a `Disconnected` event if an initial connection was never made.

And I think there may be some Coroutine magic happening that I'm not catching in the event of a successful connection, but if `spawnConnection` is able to connect to a peripheral, it will (I think) launch a new coroutine that eventually calls `onEvent?.invoke(Event.Connected(gatt))` then return `true`. And it looks like after `spawnConnection` returns `true`, a `Disconnected` event is fired, which also doesn't seem right.

Reply:
spawnConnection() returns true if the previous connection attempt was successful, IE after the connection terminates the previous connection was successful. If we do manage to connect, we don't actually return from spawnConnection until disconnection occurs.

```
                    try {
                        coroutineScope {
                            gatt.onCharacteristicChanged
                                .onEach(_onCharacteristicChanged::send)
                                .launchIn(this, start = UNDISPATCHED)
                            _gatt = gatt
                            onEvent?.invoke(Event.Connected(gatt))
                            _state.value = Connected
                        }
                    } finally {
```

We're locked in that coroutineScope block until something terminates the connection and pushes us out to the finally block. (I had to look this up to answer this, so good question)